File: Исходники/Глава 2. Часть 1/Разделяй и властвуй/4. Подсчёт инверсий в массиве/C++/Олимпиадный/CountInv.cpp
#include <vector>
#include <iostream>


std::vector<std::vector<int>> split_array_inv(std::vector<int>& array) {
    std::vector<std::vector<int>> single_array;
    for (int i : array) {
        single_array.push_back({ i });
    }
    return single_array;
}


std::pair<std::vector<int>, int> sort_arrays_inv(std::vector<int>& array_1, std::vector<int>& array_2) {
    std::vector<int> first = array_1;
    std::vector<int> second = array_2;
    int i = 0;
    int j = 0;
    int cnt_inv = 0;
    std::vector<int> new_array;
    while (i < first.size() and j < second.size()) {
        if (first[i] < second[j]) {
            new_array.push_back(first[i]);
            i++;
        }
        else {
            new_array.push_back(second[j]);
            cnt_inv += (first.size() - i);
            j++;
        }
    }
    if (i < first.size() or j < second.size()) {
        new_array.insert(new_array.end(), first.begin() + i, first.end());
        new_array.insert(new_array.end(), second.begin() + j, second.end());
    }
    return std::make_pair(new_array, cnt_inv);
}


std::pair<std::vector<int>, int> merge_inv(std::vector<std::vector<int>>& array, int cnt) {
    std::vector<std::vector<int>> final_array;
    if (array.size() % 2 > 0) {
        array.push_back({});
    }
    for (int i = 0; i < array.size() - 1; i += 2) {
        std::pair<std::vector<int>, int> res = sort_arrays_inv(array[i], array[i + 1]);
        cnt += res.second;
        final_array.push_back(res.first);
    }
    if (final_array.size() == 1) {
        return std::make_pair(final_array[0], cnt);
    }
    return merge_inv(final_array, cnt);
}


int merge_sort_inv(std::vector<int>& array) {
    std::vector<std::vector<int>> single_array = split_array_inv(array);
    return merge_inv(single_array, 0).second;
}

File: Исходники/Глава 2. Часть 1/Разделяй и властвуй/4. Подсчёт инверсий в массиве/Python/CountInv.py

def array_in_half(array):
    if len(array) > 1:
        middle = len(array) // 2
        return [array[:middle], array[middle:]]
    return array


def split_array(array):
    if len(array) == 1:
        return [array]
    first, second = array_in_half(array)
    return split_array(first) + split_array(second)


def sort_merge(arrays):
    first, second = arrays
    i = j = cnt_inv = 0
    new_array = []
    while i < len(first) and j < len(second):
        if first[i] < second[j]:
            new_array.append(first[i])
            i += 1
        else:
            new_array.append(second[j])
            cnt_inv += (len(first) - i)
            j += 1
    if i < len(first) or j < len(second):
        new_array += first[i:len(first)]
        new_array += second[j:len(second)]
    return new_array, cnt_inv


def merge(array, cnt=0):
    final_array = []
    if len(array) % 2 > 0:
        array.append([])
    for i in range(0, len(array)-1, 2):
        curr, c = sort_merge([array[i], array[i + 1]])
        cnt += c
        final_array.append(curr)
    if len(final_array) == 1:
        return final_array[0], cnt
    return merge(final_array, cnt)


def count_inv(array):
    array = split_array(array)
    return merge(array)[1]





